cloud/core/core.py

```diff
@@ -0,0 +1,130 @@
+import enum
+import heapq
+import os
+import time
+from typing import Optional
+
+import faiss
+import retro_pytorch
+import transformers
+
+from fastapi import FastAPI
+from pydantic import BaseModel
+
+# The hardcoded path to docs.
+DOCS_PATH = os.environ["DOCS_PATH"]
+
+def extractParagraphsHackily(path, name):
+    "Hackily extract paragraphs from Markdown."
+
+    paras = []
+    lines = []
+    with open(os.path.join(path, name)) as handle:
+        for line in handle.read().split("\n"):
+            # Markdown headers, ReST headers
+            if line.startswith(("#", "=")):
+                para = " ".join(lines).strip()
+                if para: paras.append(para)
+                lines = []
+            else: lines.append(line)
+    return paras
+
+ctxs = {name: extractParagraphsHackily(DOCS_PATH, name)
+        for name in os.listdir(DOCS_PATH)}
+print("Loaded:", len(ctxs), "documents,",
+      sum(len(p) for p in ctxs.values()), "paragraphs")
+
+def makeHF():
+    return transformers.pipeline("question-answering",
+    model="deepset/tinyroberta-squad2")
+
+def makeRETRO():
+    return retro_pytorch.RETRO(
```

Comment:
With your design, I believe `retro` should be located in the `/origin/workloads/retroformer`. Hopefully, there is a stateless API to provide service.

Connecting this to ingress is a different problem, which I will solve.